### ISSUE TITLE
Resource groups split by path recognition

### ISSUE THREAD

**Issue by cybrox** [rhazdon/blue_bird#47]
First of all, thanks for maintaining `blue_bird`, it certainly makes my life easier! However, I am struggling with a - probably very minor - issue.

I have normal RESTful controllers for exposing certain resources. I specify `apigroup` at the top of the controller, which works fine. However, it seems that the `resource` parameter inside the `api` declaration is not respected or resolving resource groups by path is prioritized?

With my controller having the following paths
```
GET /configurations
GET /configurations/:id
POST /configurations
DELETE /configurations/:id
```
I end up with this documentation:
![img](https://i.imgur.com/e7NUNTF.png)

How can I get blue_bird to actually bundle all those into a single resource? I did try to specify `resource` in every `api` declaration, but it doesn't seem to care about that.

For reference, I'm on
```
Erlang/OTP 20 [erts-9.1.2] [source] [64-bit] 
Elixir 1.6.0-dev (4b930b0) (compiled with OTP 20)
```

**Comment by woylie** [comment]
Hi there,

the `apigroup` macro is used to describe the resource group, which is one step higher in the hierarchy than a resource section. By default, BlueBird uses the module name of the controller as a group name, and with `apigroup`, you can overwrite the name and optionally add a description.

Currently, it is not possible to add a resource name. We tried removing that level of organization altogether (or as you put it, to bundle all those routes into a single resource), but the api blueprint specification requires resource sections to be grouped by url template.

According to the [specification](https://apiblueprint.org/documentation/specification.html#def-resource-section), it is possible to assign a name identifier to a resource section, but we haven't found a good way to do this in BlueBird yet. We'd have to specify a mapping from url to name identifier somewhere, which would probably require another macro or function. Or we'd have to add a third parameter to the `apigroup` macro for the mapping.

So in summary, it is not possible to put multiple endpoints into the same resource section, because the specification doesn't allow this.

**Comment by cybrox** [comment]
Hi,

Thanks a lot for the clarification.
I can live with that, I guess I'll just find a way to insert resource titles post-scan to make the end result look a bit nicer but it really is a minor thing. :) 

Just to be sure, though, I have seen this solved with the following kind of blueprint before. Is that actually not following specifications?
```
## My Resource Name [/myresource{id}}
### Get all [GET /myresource]
### Get single [GET]
### Create new [POST /myresource]
```

There is an example in the docs (`Section Basics > Action Sections`) that appears to do the same as far as I can see.
``` 
# Blog Posts [/posts{?limit}]
## Retrieve Blog Posts [GET]
## Create a Post [POST]
## Delete a Post [DELETE /posts/{id}]
```

Even if this is correct according to the specification, I assume there is no way to generate this with `blue_bird` at the moment, right?

**Comment by woylie** [comment]
Right now, it is not possible to configure BlueBird differently. But you're right, the specification allows some different ways to handle this. I've changed one of our test examples into this:

```
# Group Airplanes

## Airplanes [/airplanes]

### Get all airplanes [GET]

+ Response 204

### Create airplane [POST]

+ Response 204

### Delete airplane [DELETE /airplanes/{id}]

+ Response 204

### Get single airplane [GET /airplanes/{id}]

+ Response 204

### Update airplane [PUT /airplanes/{id}]

+ Response 204
```

And this is the aglio output:

![bildschirmfoto 2017-12-19 um 11 17 29](https://user-images.githubusercontent.com/13847569/34152352-9918e32a-e4ae-11e7-8897-ef821844838c.png)

Here I used the same name for the resource group and the resource section. If you don't specify a resource section name, Aglio will display `Resource` instead, so you'll end up with either Airplanes-Airplanes or Airplane-Resource.

Alternatively, it would be possible to get rid of the resource groups altogether:

```
# Airplanes [/airplanes]

## Get all airplanes [GET]

+ Response 204

## Create airplane [POST]

+ Response 204

## Delete airplane [DELETE /airplanes/{id}]

+ Response 204

## Get single airplane [GET /airplanes/{id}]

+ Response 204

## Update airplane [PUT /airplanes/{id}]

+ Response 204

# Cats [/cats]

## GET

+ Response 204

## POST

+ Response 204

## Delete cat [DELETE /cats/{id}]

+ Response 204

## Get single cat [GET /cats/{id}]

+ Response 204

## Update cat [PUT /cats/{id}]

+ Response 204
```

This results in:

![bildschirmfoto 2017-12-19 um 11 33 34](https://user-images.githubusercontent.com/13847569/34152915-7902f01a-e4b0-11e7-8075-1a3dfd4288e0.png)

Note that the resource group is unnamed here.

There is one issue with both versions though: If you don't group resource sections by URI template, you're effectively making a route identifier mandatory, because then the URI template has to defined in the action section header, and the action section can only be defined in one of the following ways:

```
## <HTTP request method>

-- or --

## <identifier> [<HTTP request method>]

-- or --

## <identifier> [<HTTP request method> <URI template>]
```

**Comment by cybrox** [comment]
Oh, that makes sense. Thanks a lot for elaborating, that really helped me create an adaption for my use case. Personally, I don't see anything wrong with having a URI template for every action (considering the standard allows it) so I forked the repo and tinkered around with it a bit.

In case anyone is reading this and is interested in getting the same result, what I did was:

**1:** Add the URI template to every route header ([147ded6](https://github.com/cybrox/blue_bird/commit/147ded6c41ed3efcb21388c494801b83a9b787a4) and [858b48e](https://github.com/cybrox/blue_bird/commit/858b48e9f0fd014cca171adc1c70c8aa3573f5f2))

**2:** Fetch the resource name from our URL format. Unlike **1**, this is very specific to my needs and should definitely not be blindly copied ([cb77893](https://github.com/cybrox/blue_bird/commit/cb77893ca40f0328f3b429a08cc5e40c6a71c8cb))

**3:** Since I use an `apidocs` binstub for (re-) generating docs anyways, manually remove the `resource/:id` header from the file after generating it.
```
# Remove between-resources header for / and /:id resources
sed -i '' '/^##[^#].*[{]id[}]/d' doc/api/api.apib
```

Boom!
![https://i.imgur.com/hIUQ8JL.png](https://i.imgur.com/hIUQ8JL.png)

I think the changes from **1** are actually something that could be applied to `blue_bird` itself, as I don't see any downsides, however, you definitely have more insight as to where that could potentially cause problems.

(Also thanks for the hint that just dropping the headers would have had that kind of outcome, I did not consider that at all. However, since I have one or two controllers that bundle multiple resources, I think just removing the unnecessary header is cleaner than removing them all)

Thanks again, I'm happy with this solution and I'll be happy to drop it for an updated version of `blue_bird` if it happens to change resource section handling again in the future. 👍

**Comment by woylie** [comment]
I'm glad you found a solution. Your fork does not produce valid apib files, though – I'm getting a lot of `the resource '/routeX' is already defined (warning code 2)` errors from aglio.

**Comment by cybrox** [comment]
Thanks for testing. When testing with the example `.apib` in this repo's `/priv` and the API of the project I'm working on, I did not encounter any `code 2` errors. What data did you attempt to format with the forked version?

**Comment by woylie** [comment]
There are some test files in `test/support/examples`. Each file includes an `ApiDoc` struct, the expected apib output and the expected Swagger map. The task `mix bird.gen.examples` creates apib and swagger files from all of those ApiDocs and then uses aglio to render html files (from apib only). You'll find the files in `priv/static/docs`.

The mix task only writes the files, but it does not test the output. That is only done in the regular.

There are still some issues with the original BlueBird output, but the error above is not among them.
